[고급 레이아웃 컴포넌트] 야미(이다인) 미션 제출합니다

feb-dain opened the pull request:
# 🎯 2단계. 고급 레이아웃 컴포넌트

### 안녕하세요 우스! 2단계도 잘 부탁드립니다 ㅎㅎ
이름이 이해가 안 되거나 사용하기 불편하다면 편하게 말씀해주세요~

[🚀 storybook](https://65054a5accdaa5b7c9142b51-yxsqzijqoo.chromatic.com/?path=/docs/instruction--docs)

## 공통
- [x] 의미있는 커밋 메시지 작성
- [x] PR에 관련 없는 변경 사항이 포함 유무 확인

## 컴포넌트 요구사항
- [x] 선택한 컴포넌트의 기능 구현
- [x] npm 배포
- [x] README.md에 코드 저자로서 특히 고민한 부분과 의도 설명
- [ ] 테스트 코드를 작성하였나요? (선택 사항)

## 📌 구현한 내용 설명

재사용 가능하고, 이용하기 쉬운 탭 컴포넌트를 구현하려고 노력했습니다!

## 📸 스크린샷 (선택 사항)

(필요하다면 구현한 내용을 보여주는 스크린샷 첨부.)

## 🔗 참고 링크 (선택 사항)

(유용했던 참고 자료나 레퍼런스가 있다면 링크 첨부)

---

# 📝 리뷰어를 위한 가이드라인

이번 미션의 핵심 경험은 'CSS'를 더 잘 활용하는 것입니다. 그래서 코드 리뷰 진행 시, React를 이용한 기능적인 부분보다는 아래와 같은 가이드라인을 '참고'하여, CSS의 개선을 고려한 피드백을 남겨주세요.

### 레이아웃 컴포넌트가 여러 환경에서도 잘 보이는지 확인해 주세요
- [ ] 반응형: 레이아웃이 다양한 화면 크기에 적응하는지 확인해 주세요.
- [ ] 크로스 브라우징: 여러 브라우저(크롬, 사파리, 엣지 등)에서 동작하는지 확인해 주세요.

### 레이아웃 컴포넌트를 사용하는 사용자로서, 잘 이해하고 활용할 수 있었는지 피드백해 주세요
- [ ] 가독성: README나 사용법을 보고, 저자의 의도가 잘 이해되는지 독자의 입장에서 확인해 주세요.
- [ ] 사용성: 레이아웃 컴포넌트를 사용하는 입장에서 쉽게 사용할 수 있었는지 확인해 주세요.

### 스타일 코드가 네이밍, 재사용성 등을 고려해 잘 작성되어 있는지 확인해 주세요
- [ ] 의미론적 네이밍: 해당 CSS 클래스나 ID의 이름이 의미론적으로 적절한지 확인해 주세요.
- [ ] 재사용성: 중복되거나 불필요한 스타일 규칙이 있는지 살펴보세요.

feb-dain committed:
refactor: children의 타입 변경

feb-dain committed:
refactor: 테스트용으로 넣어놨던 Grid 컴포넌트 제거

feb-dain committed:
feat: `Tab` 컴포넌트 구현

feb-dain committed:
refactor: `Tab` 컴포넌트 export

feb-dain committed:
chore: 자동 배포 브랜치 변경

feb-dain committed:
fix: cra 배포 오류 개선

feb-dain committed:
refactor: Menu에 button 속성 추가

feb-dain committed:
chore: 버전 업데이트

feb-dain committed:
fix: 배포 오류 개선

feb-dain committed:
design: `Tab` 디자인 개선

feb-dain committed:
refactor: 웹 접근성 개선

feb-dain committed:
chore: 버전 업데이트

feb-dain committed:
docs: 스토리북 링크 업데이트

Gilpop8663 commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "yummy-design-component",
-  "version": "0.1.6",
+  "version": "0.2.2",
```
열심히 하셨군요 👍👍👍

Gilpop8663 commented on src/lib/common/Tab/Content.tsx:
```diff
@@ -0,0 +1,29 @@
+  if (activeTab !== `menu-${index}`) return <></>;
+
+  return (
+    <FlexItem alignSelf={position} role="tabpanel">
```
우와 웹 접근성이 눈에 띄네여 💯💯💯

Gilpop8663 commented on src/lib/common/Tab/Menus.tsx:
```diff
@@ -0,0 +1,71 @@
+import { TabContext } from './Tab';
+
+export interface MenusProps extends FlexContainerProps {
+  /** 선택된 탭 메뉴를 강조하는 색깔 변경 가능
+   * @default #000
+   */
+  highlightColor?: string;
+  /** 선택된 탭 메뉴 아래에 선 제거 가능
+   * @default false
+   */
+  noUnderline?: boolean;
```
주석으로 인해서 라이브러리를 사용할 때 사용자가 쉽게 prop이 어떤 동작을 하는지 유추할 수 있어서 좋다고 생각이 들어요 🔥🔥🔥

Gilpop8663 commented on src/lib/common/Tab/Menus.tsx:
```diff
@@ -0,0 +1,71 @@
+
+  return (
+    <FlexContainer
+      role="tablist"
```
여기도 웹 접근성을 챙기셨네요 🔥

Gilpop8663 commented on vite.config.ts:
```diff
@@ -1,8 +1,9 @@
 import react from '@vitejs/plugin-react';
 import { defineConfig } from 'vite';
+import dts from 'vite-plugin-dts';
```
vite-plugin-dts로 설정을 하셨는데 설정에 오류가 났거나 npm으로 publish가 안된 것 같아요

설치한 패키지에 index.d.ts가 존재하지 않아서 자동완성이 안되네요 🤕

![image](https://github.com/woowacourse/layout-component/assets/80146176/28bb87ad-3014-41ec-99d6-70cff3e1c864)

![image](https://github.com/woowacourse/layout-component/assets/80146176/1b479e37-92b1-4765-8b04-41a5ff406ad6)

Gilpop8663 reviewed (changes requested):
안녕하세요 야미 잘 부탁드립니당

### 리뷰는   "yummy-design-component": "^0.2.2" 기준으로 작성했습니다


## 여러 환경에서 잘 보이는가?

반응형: 휴대폰, 탭, 데스크탑에서 동등하게 작동하는 것을 확인했습니다 🌊🌊🌊
크로스 브라우징: 크롬, 브레이브, 엣지, 파이어폭스에서 동등하게 작동하는 것을 확인했습니다 🌊🌊🌊



##  문서가 가독성 있게 쓰여있는가? 

스토리북에 사용 예제가 있어서 이해하는데 큰 도움이 되었습니다. 그리고 탭 컴포넌트에 대한필수 설명이 잘 되어 있어서 좋았어요. 근데 궁금한점은 스토리북 Tab, Menu에서 layout Props가 존재하는 것이 궁금하였습니다 따로 컨트롤 해보았는데 변하는 것이 없더라구요. layout을 포함하신 이유가 궁금해요

![image](https://github.com/woowacourse/layout-component/assets/80146176/c467fbae-59a0-4f88-9a56-0dc4be771a34)


## 레이아웃 컴포넌트를 사용하는 입장에서 쉽게 사용할 수 있었는지 확인해 주세요. 고려한 사용자에게 발생할 수 있는 상황에 따른 props를 고려했는가?

저의 경우 야미의 탭 컴포넌트가 사용 방법이 명확해서 쉽게 사용이 가능하다는 생각이 들엇습니다. 그리고 현재 있는 props 원하는 디자인의 탭을 동작 시킬 수 있다는 생각이 들었어요 👍👍👍


##  ‘유연’하게 재사용 하기 좋은가? 


1. 사용자 입장에서 웹페이지의 컬러에 맞게 Tab Component의 색상을 변경할 수 있는 속성이 있어서 너무 좋았습니다

2. 라벨 텍스트가 매우 긴 경우에는 Menu의 최대 길이를 설정할 수 있게 한 뒤 text-overflow: ellipsis; 을 적용해서 ...으로 보여주는 것은 어떨까요? 그리고 마우스를 올리면 모든 텍스트가 보이는 식으로요

### 현재는 메뉴의 라벨 텍스트의 길이 상관없이 모두 보여주고 있습니다

![image](https://github.com/woowacourse/layout-component/assets/80146176/652f2573-5d49-4165-98ad-180708840bcf)

### 예시로 들은 이미지 

![image](https://github.com/woowacourse/layout-component/assets/80146176/4f1fe04d-361a-41e7-92cb-14ed0da1de2b)


https://developer.mozilla.org/en-US/docs/Web/CSS/text-overflow



3. 가로 외에 세로형으로도 지원을 해줘서 범용적으로 사용할 수 있는 Tab 패키지라고 생각이 들었어요 💯💯💯


4. 탭의 수가 많은 경우에는 Tab.Menus의 width 혹은 height 값만큼만 보여주고, Tab.Menus 내부에서 스크롤 되도록 overflow:scroll을 해주는 것은 어떠세요?

### 현재는 탭이 많을 때 페이지 전체에 스크롤이 생기는 것 같아요

![image](https://github.com/woowacourse/layout-component/assets/80146176/6b3661d6-4b8e-464a-8799-54722f5c38e6)

5. 패키지를 확인하기 위해 vite로 새로 설치해서 사용했는데요. reset.css가 적용되지 않아서 그런지 스토리북의 코드를 똑같이 사용했을 때 CSS 디자인이 미묘하게 다른 점이 있어서 제보합니다

button에 border-radius 8px, outline 등이  적용되어서 스토리북과는 다르게 표시되는 것 같아요

https://github.com/woowacourse/layout-component/assets/80146176/4494101c-2642-4135-8109-f9c824d07871


--------

웹 접근성을 챙긴 부분과 꼼꼼한 문서화 하신 부분 너무 멋있어요 코드 보면서 많이 배우고 있습니다 탭 컴포넌트를 사용하기 쉽게 만드신 부분이 좋은 컴포넌트라고 생각 들었어요. 

좋았던 점과 코드를 보며 궁금한 점 리뷰 남겼습니다 🔥🔥🔥

feb-dain committed:
fix: 배포 문제 개선

feb-dain committed:
refactor: 필요없는 코드 제거

feb-dain committed:
test: 여러 탭 쓰는 경우를 보여주는 미리보기 추가

feb-dain committed:
refactor: 필요없는 코드 제거 및 Menu 스토리 추가

feb-dain committed:
refactor: `hexToLighterColor` 유틸 개선

feb-dain committed:
test: layout 속성이 적용된 Tab 예시 추가 및 스토리 설명 개선

feb-dain committed:
chore: `react-icons` 설치

feb-dain committed:
feat: `Menu` 컴포넌트에 iconPosition 속성 추가

feb-dain committed:
refactor: style 분리 및 파일 이름 변경

feb-dain committed:
test: iconPosition 설명 추가

feb-dain committed:
test: iconPosition 테스트할 수 있는 기본 탭 스토리에 추가

feb-dain committed:
test: `InitialIndex` 스토리 잘못된 부분 개선

feb-dain committed:
refactor: 웹 접근성 개선

feb-dain committed:
test: `Tab` 관련 스토리에 css 설명 추가

feb-dain committed:
feat: Menus에 `lineClamp` 속성 추가

feb-dain committed:
design: 해상도마다 양 옆에도 box-shadow가 보이는 현상 개선

feb-dain committed:
design: 내용 영역은 그대로 있고 탭 메뉴들만 스크롤 되도록 변경

- 기존: 탭 메뉴들의 너비가 부모 컨테이너 보다 클 경우, 탭 전체가 스크롤 됨
- 수정 후: 내용 영역은 그대로 있고, 탭 메뉴들 영역만 스크롤

feb-dain committed:
design: reset css 적용 및 파일 이동/경로 변경

feb-dain committed:
chore: 버전 업데이트

feb-dain committed:
docs: 스토리북 링크 업데이트

feb-dain committed:
refactor: Menus에서 `layout` 속성 제거

feb-dain committed:
docs: 스토리북 링크 업데이트

feb-dain commented:
안녕하세요 우스! 정성스러운 리뷰 감사합니다 ㅎㅎ

<a href="https://65054a5accdaa5b7c9142b51-fcuwgvpksi.chromatic.com/?path=/docs/instruction--docs">🚀 스토리북</a> 

> 스토리북에 사용 예제가 있어서 이해하는데 큰 도움이 되었습니다. 그리고 탭 컴포넌트에 대한필수 설명이 잘 되어 있어서 좋았어요. 근데 궁금한점은 스토리북 Tab, Menu에서 layout Props가 존재하는 것이 궁금하였습니다 따로 컨트롤 해보았는데 변하는 것이 없더라구요. layout을 포함하신 이유가 궁금해요

layout 속성에 맞춰서 컴포넌트의 위치가 변합니다! 그런데 스토리북 상에서 아래와 같은 이유 때문에 변화가 안 보였던 것 같아요😂
 1. Tab의 높이가 탭 메뉴들과 내용에 맞춰져 있기 때문
 2. vertical이 적용되지 않은 Menus는 너비가 100%이기 때문
 3. Content의 alignSelf 속성 기본값이 center이기 때문

충분히 오해할 수 있는 상황이라고 생각해 스토리북에 예제 스토리를 추가했습니다!
`Tab With Layout Prop`에서 layout 속성을 바꿔보면 컴포넌트의 변화가 눈에 보입니다 ㅎㅎ 

+ Menus에도 layout 속성으로 변화를 줄 수 있지만, 변화를 보기 위해서 따로 설정하는 게 번거롭기도 하고 굳이 필요없을 것 같아 제거 했습니다!👍🏻

> 라벨 텍스트가 매우 긴 경우에는 Menu의 최대 길이를 설정할 수 있게 한 뒤 text-overflow: ellipsis; 을 적용해서 ...으로 보여주는 것은 어떨까요? 그리고 마우스를 올리면 모든 텍스트가 보이는 식으로요

👍🏻좋은 생각인 것 같습니다! `lineClamp`라는 속성을 추가해 true일 경우, 텍스트가 여러 줄이 되면 ...로 보여지도록 했습니다!

> 탭의 수가 많은 경우에는 Tab.Menus의 width 혹은 height 값만큼만 보여주고, Tab.Menus 내부에서 스크롤 되도록 overflow:scroll을 해주는 것은 어떠세요?

 그렇게 되는 게 더 좋을 것 같아 반영했습니다 🚀

> 패키지를 확인하기 위해 vite로 새로 설치해서 사용했는데요. reset.css가 적용되지 않아서 그런지 스토리북의 코드를 똑같이 사용했을 때 CSS 디자인이 미묘하게 다른 점이 있어서 제보합니다

 reset css를 적용시켜 동일한 디자인으로 보이도록 만들었습니다! 스토리북에만 font를 적용시켜놔서 font style은 다를 수 있습니다😄

Gilpop8663 commented on src/lib/common/Tab/Menu.tsx:
```diff
@@ -0,0 +1,46 @@
+  /** 탭 메뉴 아이콘 */
+  icon?: ReactNode;
+  /** 탭 메뉴 아이콘 위치, 아이콘이 있을 때만 동작
+   * @default 'left'
```
이렇게 default 값을 알려주는 것 너무 좋네요! 배워갑니당 👍👍👍

Gilpop8663 reviewed (changes requested):
안녕하세요 야미

스토리북 문서의 정보의 종류가 더 많아지고, 설명도 더 친절해진 것을 느낄 수 있었어요 💯💯💯

그리고 탭에서 lineClamp ... 기능이 잘 동작하는 것을 확인할 수 있었어요 👍👍👍

## 버그 제보 (CSS)

그렇지만 아직 reset.css가 제대로 먹히지 않는 것을 확인할 수 있었는데요

### 적용 전

![image](https://github.com/woowacourse/layout-component/assets/80146176/46268c8f-b2af-4b8b-adaa-9a4746209b69)


사용하는 곳에서  아래의 CSS 속성을 따로 추가하고 난 뒤 의도한대로 보이게 되었어요  따로 의도하신 부분이 아니라면 따로 사용하는 측에서 CSS 속성을 건드리지 않고 적용되면 좋겠다는 생각이 들었습니다

```css
button:focus,
button:active {
  outline: none;
}

button {
  border-radius: 0;
}
```

### 적용 후

![image](https://github.com/woowacourse/layout-component/assets/80146176/6607bd03-2623-4d17-b03a-6b1890371939)


## 버그 제보 (Menus 속성 Props들이 자동 완성에서 다 안 나타나요)

스토리북에서 보면 width, height ,mx 등 여러 속성이 있고, 실제로 자동 완성은 안되더라도 직접 입력하면 잘 동작이 되는데요. Tab.Menus에서 모든 Props들이 자동 완성이 되면 더 좋겠다는 생각이 들었습니다.


![image](https://github.com/woowacourse/layout-component/assets/80146176/1f13b6ab-6f94-4a08-a617-7556091dc278)




 "yummy-design-component": "^1.0.2" 버전 기준으로  리뷰 달았습니다

코멘트 보시고 한번 확인해주시면 감사하겠습니다

feb-dain committed:
design: button 관련 속성 reset하는 css 추가

feb-dain committed:
fix: `Menus` 자동 완성 안 되는 오류 개선

feb-dain commented:
`Omit` 쓰면 자동 완성이 잘 안 되는 걸 처음 알았네요😂 
reset.css에 우스가 알려준 css 추가하고, Menus 자동 완성 되도록 변경했습니다! 

제가 크롬에서 테스트 했을 때는 border-radius가 없었는데, 혹시 어떤 브라우저 사용했는지 알려주실 수 있나요?

Gilpop8663 reviewed (approved):
Menus에서 자동 완성 되는 부분 잘 확인했습니다~

그리고 reset.css 문제가 아니라 제가 실행하던 환경에서 index.css에서 button에 border-radius 속성을 주고 있더라구요. 제대로 확인 안햇던 점 죄송해요 ㅠㅠ

Tabs 컴포넌트 작동 잘 되는 것 확인하고, 제안하였던 부분들 코드 반영된 것 확인 후 어프로브 했습니다~~ 화이팅이에요~!

-------

### 프로젝트 파일에 자동으로 적용되던  index.css ;;

```tsx
import React from 'react'
import ReactDOM from 'react-dom/client'
import App from './App.tsx'
import './index.css'

ReactDOM.createRoot(document.getElementById('root')!).render(
  <React.StrictMode>
    <App />
  </React.StrictMode>,
)

```

```css
button {
  border-radius: 8px;
  border: 1px solid transparent;
  padding: 0.6em 1.2em;
  font-size: 1em;
  font-weight: 500;
  font-family: inherit;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.25s;
}
button:hover {
  border-color: #646cff;
}
button:focus,
button:focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

@media (prefers-color-scheme: light) {
  :root {
    color: #213547;
    background-color: #ffffff;
  }
  a:hover {
    color: #747bff;
  }
  button {
    background-color: #f9f9f9;
  }
}

```
